### ISSUE TITLE
Tracing DataReader timings?

### ISSUE THREAD

**Issue by yallie** [linq2db/linq2db#1355]
Hello,

I'm planning to update my project from BLToolkit to the latest version of Linq2db. 
My project relies on some BLToolkit internals for tracing SQL queries.
Apparently they are not relevant for Linq2db :)

I've implemented custom DataReader class to measure query execution timings:
* from ICommand.ExecuteReader to IDataReader.Read → query execution time
* from IDataReader.Read to IDataReader.Dispose → data fetching time

I used to use custom data providers with their own MappingSchema classes and
override CreateDataReaderMapper to wrap the original DataReader with my wrapper.
That worked just fine for my BLToolkit-based application.

What can I do in Linq2db to achieve similar functionality?
I see that DataConnection has this handy OnTraceConnection callback.
That's cool, but I don't seem to be able to track DataReader's lifetime with that.
Can I plug into the data provider pipeline to wrap the original DataReader with my own class?
Or is there any better option?

Thanks :)

**Comment by sdanyliv** [comment]
@yallie, `OnTraceConnection` gives all needed information

>from ICommand.ExecuteReader to IDataReader.Read → query execution time

TraceInfoStep.AfterExecute get info.ExecutionTime

>from IDataReader.Read to IDataReader.Dispose → data fetching time

TraceInfoStep.Completed get info.ExecutionTime

>but I don't seem to be able to track DataReader's lifetime with that.

Do you still needs injecting in this functionality?

**Comment by yallie** [comment]
Oh, that sounds great!
If TraceInfoStep.Completed gives the fetch timings, that's all I need :)
The only reason I would wrap the DataReader is to measure these timings.

But looking at the sources, I see that Completed step is only invoked by QueryRunner.
So I assume, it won't be triggered when I read the DataReader myself, right?
I don't seem to be getting the Completed step when doing something like that:

```c#
using (var db = new DataConnection())
{
	DataConnection.TurnTraceSwitchOn();
	db.OnTraceConnection = (e) =>
	{
		Console.WriteLine("Step: {0}, Time: {1}", e.TraceInfoStep, e.ExecutionTime);
	};

	using (var reader = db.ExecuteReader("select * from my_table"))
	{
		var query = reader.Query<MyTable>(); // mapped to my_table
		foreach (var record in query)
		{
			Console.WriteLine($"Record: {record}");
		}
	}
}
```

Console log:

```
Step: BeforeExecute, Time:
Step: AfterExecute, Time: 00:00:00.0189886
Record: ID = 1, Name = One
Record: ID = 2, Name = Two
Record: ID = 3, Name = Three
```

**Comment by yallie** [comment]
Perhaps we could make LinqToDB.Data.DataReader itself invoke the TraceInfoStep.Completed step?

**Comment by sdanyliv** [comment]
Yes there is no wrapping around that because for timing for now is responsible user. 
Just for confirmation, why do you need such approach? Linq2db has fastest mapper ever. Just use 
```cs
foreach (var r in db.Query<MyTable>(...))
{
    Console.WriteLine($"Record: {record}");
}
```

**Comment by yallie** [comment]
>Linq2db has fastest mapper ever. Just use
>foreach (var r in db.Query<MyTable>(...))

That was just a sample :)

My app is 95% LINQ, but a few dozen queries are plain SQL.
So sometimes I'm forced to use the DataReader.

**Comment by sdanyliv** [comment]
Also, you do not need to write SQL by yourself, linq2db supports almost everything for expressing your needs. If you feel that we cant’t do that, post SQL sample and you will be surprised.

**Comment by yallie** [comment]
Good point, but there are reasons we couldn't avoid plain SQL :)

1. The application is large, and we'd like to avoid rewriting existing SQL queries in LINQ. 
2. We have huge queries generated by report generator in plain SQL.
3. Some queries use database-specific stuff like writable CTEs, group by rollup, connect by, etc.

**Comment by sdanyliv** [comment]
@yallie, PR always wellcome. It can be DataReader wrapper. For sure it should not affect performance.
